pkg/reconciler/route/traffic/rollout.go

```diff
@@ -70,3 +72,94 @@ type RevisionRollout struct {
 	// target percentage.
 	Percent int `json:"percent"`
 }
+
+// Roll processes previous Rollout object and compares to the current
+// rollout state. If there is different, Merge will start or stop the rollout
+// and update `RevisionRollout` objects accoringly.
+// Roll returns true if any changes have been made.
+func (cur *Rollout) Roll(prev *Rollout) bool {
+	if prev == nil {
+		return false
+	}
+
+	// The algorithm below is simplest, but probably not the most performant.
+	// Optimize in the later passes.
+	// Map by tag.
+	ccfg, pcfg := map[string][]*ConfigurationRollout{}, map[string][]*ConfigurationRollout{}
+	for i := range cur.Configurations {
+		cfg := &cur.Configurations[i]
+		ccfg[cfg.Tag] = append(ccfg[cfg.Tag], cfg)
+	}
+	for i := range prev.Configurations {
+		cfg := &prev.Configurations[i]
+		pcfg[cfg.Tag] = append(pcfg[cfg.Tag], cfg)
+	}
+	ret := false
+	for t, cfgs := range ccfg {
+		pcfgs, ok := pcfg[t]
```

Comment:
I find these variable names *very* confusing 😞

Reply:
Oook, I can do that.

Reply:
Should be better now.